database/src/ops/block.rs

```diff
@@ -1,89 +1,537 @@
+    let (block_height, block_hash) = {
+        let (block_height, block_info) = tables.block_infos_mut().pop_last()?;
+        (block_height, block_info.block_hash)
+    };
+
+    // Block heights.
+    tables.block_heights_mut().delete(&block_hash)?;
+
+    // Block blobs.
+    // We deserialize the block blob into a `Block`, such
+    // that we can remove the associated transactions later.
+    let block_blob = tables.block_blobs_mut().take(&block_height)?.0;
+    let block = Block::read(&mut block_blob.as_slice())?;
+
+    // Transaction & Outputs.
+    for tx_hash in block.txs {
+        let (tx_id, tx_blob) = remove_tx(&tx_hash, tables)?;
+        let tx = Transaction::read(&mut tx_blob.0.as_slice())?;
+
+        // Is this a miner transaction?
+        let mut miner_tx = false;
+        for inputs in &tx.prefix.inputs {
+            match inputs {
+                // Key images.
+                Input::ToKey { key_image, .. } => {
+                    add_key_image(tables.key_images_mut(), key_image.compress().as_bytes())?;
+                }
+                // This is a miner transaction, set it for later use.
+                Input::Gen(_) => miner_tx = true,
+            }
+        }
+
+        // Remove each output in the transaction.
+        for output in tx.prefix.outputs {
+            // Outputs with clear amounts.
+            if let Some(amount) = output.amount {
+                // RingCT miner outputs.
+                if miner_tx && tx.prefix.version == 2 {
+                    let amount_index = get_rct_num_outputs(tables.rct_outputs_mut())?;
+                    remove_rct_output(&amount_index, tables.rct_outputs_mut())?;
+                // Pre-RingCT outputs.
+                } else {
+                    let amount_index = tables.num_outputs_mut().take(&amount)?;
+                    remove_output(
+                        &PreRctOutputId {
+                            amount,
+                            amount_index,
+                        },
+                        tables,
+                    )?;
+                }
+            // RingCT outputs.
+            } else {
+                let amount_index = get_rct_num_outputs(tables.rct_outputs_mut())?;
+                remove_rct_output(&amount_index, tables.rct_outputs_mut())?;
+            }
+        }
+    }
+
+    Ok(block_height)
 }
 
-/// TODO
-pub fn get_block_timestamp() {
-    todo!()
+//---------------------------------------------------------------------------------------------------- `get_block_*`
+/// Retrieve a [`VerifiedBlockInformation`] from the database.
+///
+/// This extracts all the data from the database tables
+/// needed to create a full `VerifiedBlockInformation`.
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*};
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[inline]
+pub fn get_block(
+    tables: &impl Tables,
+    block_hash: BlockHash,
+) -> Result<VerifiedBlockInformation, RuntimeError> {
```

Comment:
I don't think this function will be needed FWIW

Reply:
I was just about to ping you saying that returning `VerifiedBlockInformation` is impossible within `cuprate_database` unless we pulled in `consensus/` for stuff like `pow_hash` but then realized `Response` only needs `ExtendedBlockHeader` and `monero_serai::Block`...

Should these be the only functions we have for retrieval?
```rust
fn get_block(...) -> Result<Block, RuntimeError>; // for other users
fn get_block_and_tx(...) -> Result<(Block, Vec<Transaction>), RuntimeError>;
fn get_block_extended_header(...) -> Result<ExtendedBlockHeader, RuntimeError>;
```

Reply:
we can probably get away with just `get_block_extended_header` for now. `BlockBatchInRange` was a hack for getting the RPC scanner to work so can be removed.

We will need a request for block and tx blobs but we don't need to do that now.